data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/ScanObjectWorker.java

```diff
@@ -0,0 +1,175 @@
+
+    private final List<ScanOptions> scanOptionsBuilderList;
+
+    private final S3ObjectHandler s3ObjectHandler;
+
+    public ScanObjectWorker(final S3Client s3Client,
+                            final List<ScanOptions> scanOptionsBuilderList,
+                            final S3ObjectHandler s3ObjectHandler){
+        this.s3Client = s3Client;
+        this.scanOptionsBuilderList = scanOptionsBuilderList;
+        this.s3ObjectHandler= s3ObjectHandler;
+    }
+
+    /**
+     * It will decide the s3 object parse <code>S3ObjectWorker</code> or <code>S3SelectWorker</code>
+     * based on s3 select configuration provided.
+     */
+    @Override
+    public void run() {
+        scanOptionsBuilderList.forEach(this::parseS3ScanObjects);
+    }
+
+    /**
+     * Method will parse the s3 object and write to {@link Buffer}
+     */
+    void parseS3ScanObjects(final ScanOptions scanOptions) {
+        scanIncludeObjectsValidationCheck(scanOptions);
+        scanOptions.getIncludeKeyPaths().forEach(key ->{
+            ListObjectsV2Request request = ListObjectsV2Request.builder()
+                    .bucket(scanOptions.getBucket())
+                    .prefix(key).build();
+
+            final ListObjectsV2Response s3ObjectResponse = s3Client.listObjectsV2(request);
+            s3ObjectResponse.contents().stream().map(S3Object::key).filter(s3ObjKey-> s3ObjKey.lastIndexOf(".")!=-1)
+                    .filter(s -> !scanOptions.getExcludeKeyPaths().contains(s.substring(s.lastIndexOf("."))))
+                    .forEach(s3Object ->
+                        processS3ObjectKeys(S3ObjectReference.bucketAndKey(scanOptions.getBucket(),
+                                s3Object).build(),s3ObjectHandler, scanOptions));
+        });
+    }
+
+    private static void scanIncludeObjectsValidationCheck(final ScanOptions scanOptions) {
+        if(Objects.isNull(scanOptions.getIncludeKeyPaths()) || Objects.isNull(scanOptions.getExcludeKeyPaths())){
+            throw new IllegalArgumentException("include/exclude list should not be empty in pipeline yaml configuration");
+        }
+    }
+
+    private void processS3ObjectKeys(final S3ObjectReference s3ObjectReference,
+                                     final S3ObjectHandler s3ObjectHandler,
+                                     final ScanOptions scanOptions){
+        final S3ObjectDetails s3ObjDetails = getS3ObjectDetails(s3ObjectReference);
+        final boolean isKeyMatchedBetweenTimeRange = isKeyMatchedBetweenTimeRange(s3ObjDetails.getS3ObjectLastModifiedTimestamp(),
+                scanOptions.getRange(),
+                scanOptions.getStartDateTime(),
+                scanOptions.getEndDateTime());
+        if(isKeyMatchedBetweenTimeRange && (isKeyProcessedByS3Scan(s3ObjDetails))){
+            updateKeyProcessedByS3Scan(s3ObjDetails);
+            try{
+                s3ObjectHandler.parseS3Object(s3ObjectReference,null);
+            }catch (IOException ex){
```

Comment:
Can you confirm if `s3ObjectHandler.parseS3Object` only throws IOException ? We might be missing some exception handling.

Reply:
Thanks for your review Dinu, yes it is.